### DISCUSSION TITLE
Clarification about column name case for dialect maintainers (`requires_name_normalize = True`?)

### DISCUSSION THREAD

**Discussion by fdcastel** [sqlalchemy/sqlalchemy#10781] [Usage Questions] [answered]
I just noticed Firebird dialect does not work as other dialects when returning the column names.

```python
def test_column_names_mssql():
    e = sa.create_engine(db_uri)
    with e.connect() as conn:
        q = 'SELECT 1 one, 2 TWO, 3 "three", 4 "FOUR", 5 "Five"'
        r = conn.exec_driver_sql(q)

        expected = ["one", "TWO", "three", "FOUR", "Five"]  # Sql Server preserves case even for unquoted names
        assert [c for c, *rest in r.cursor.description] == expected
        assert [c for c in r.mappings().first()] == expected  # Correct. Same as r.cursor.description

def test_column_names_postgresql():
    e = sa.create_engine(db_uri)
    with e.connect() as conn:
        q = 'SELECT 1 one, 2 TWO, 3 "three", 4 "FOUR", 5 "Five"'
        r = conn.exec_driver_sql(q)

        expected = ["one", "two", "three", "FOUR", "Five"]  # PostgreSQL uses lowercase for unquoted names
        assert [c for c, *rest in r.cursor.description] == expected
        assert [c for c in r.mappings().first()] == expected  # Correct. Same as r.cursor.description

def test_column_names_firebird():
    e = sa.create_engine(db_uri)
    with e.connect() as conn:
        q = 'SELECT 1 one, 2 TWO, 3 "three", 4 "FOUR", 5 "Five" from rdb$database'
        r = conn.exec_driver_sql(q)

        expected = ["ONE", "TWO", "three", "FOUR", "Five"]  # Firebird uses uppercase for unquoted names
        assert [c for c, *rest in r.cursor.description] == expected

        unexpected = ["one", "two", "three", "four", "Five"]
        assert [c for c in r.mappings().first()] == unexpected  # Incorrect. Column names differ from r.cursor.description
```

A deeper investigation brought me to [this line](https://github.com/sqlalchemy/sqlalchemy/blob/ca9adf3028095ff28bdef823d6f6cefc12815b3c/lib/sqlalchemy/engine/cursor.py#L673) in `cursor.py` and the realization that only [Firebird](https://github.com/pauldex/sqlalchemy-firebird/blob/main/sqlalchemy_firebird/base.py#L411) and [Oracle](https://github.com/sqlalchemy/sqlalchemy/blob/ca9adf3028095ff28bdef823d6f6cefc12815b3c/lib/sqlalchemy/dialects/oracle/base.py#L1428) dialects actually use `requires_name_normalize = True` (default is `False` for all other dialects).

My first reaction was to set `requires_name_normalize` to `False` in Firebird dialect. But then all hell broke loose since **most of tests are failing now** (because column names doesn't match anymore).

I made some attempts to implement a `normalize_name`/`denormalize_name` but they all failed.

**What should be the correct implementation for Firebird?** 

For reference: Firebird returns all column names in UPPERCASE, except when surrounded by double quotes. This can be seen in the above tests (`r.cursor.description`).

_EDIT: Added PostgreSQL and removed Pandas from sample code._

**Comment by fdcastel** [comment]
Also: Shouldn't all column names always match exactly to what `cursor.description` returns?

If yes, could we add a test for this?

**Reply by gordthompson** [reply]
FWIW, I see the same results from Firebird as I do from Oracle, namely

```python
with engine.connect() as conn:
    q = 'SELECT 1 one, 2 TWO, 3 "three", 4 "FOUR", 5 "Five" from DUAL'
    r = conn.exec_driver_sql(q)
    print(r.mappings().all())
    # [{'one': 1, 'two': 2, 'three': 3, 'four': 4, 'Five': 5}]
```

This seems to be in line with SQLAlchemy's "lower-case-means-insensitive convention" as described [here](https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.quoted_name).

**Reply by fdcastel** [reply]
Thank you @gordthompson for the documentation link and the sample code. I didn't have a Oracle server to test this.

Anyway, to the best of my understanding, the Oracle dialect is also wrong.

That `"FOUR"` should never be lowercased in this example.

What `r.cursor.description` returns to you?

**Reply by gordthompson** [reply]
```python
    crsr = conn.connection.cursor()
    crsr.execute(q)
    print([col[0] for col in crsr.description])
    # ['ONE', 'TWO', 'three', 'FOUR', 'Five']
```

**Reply by fdcastel** [reply]
Thanks. As I presumed, the driver returns the column names correctly.

So, my point is: How do I make SQLAlchemy keep the column names as intended by the original SQL query?

In other dialects this works. Except in Oracle and Firebird.

**Reply by zzzeek** [reply]
if you are looking at cursor.description, you are bypassing SQLAlchemy.   look in r.keys() and that will have the names

**Reply by fdcastel** [reply]
@zzzeek I'm looking at `cursor.description` only to assert that the problem is not coming from the driver.

Problem is: **why SQLAlchemy is changing the column names?**

Again: Shouldn't all column names always match exactly to what `cursor.description` returns?

Please look at the sample code of original post. It shows **two different behaviors** for SQL Server and Firebird dialects.

**Reply by fdcastel** [reply]
I've added PostgreSQL behavior to the sample code from the original post.

Also, I changed the code to use `r.mappings` instead of Pandas, per @gordthompson sample.

**Reply by zzzeek** [reply]
> Again: Shouldn't all column names always match exactly to what cursor.description returns?

it does in all but one case, if you consider SQLAlchemy uses all lower case to indicate case insensitive names.

the one case is if you use an uppercase-for-case-insensitive database like Firebird or Oracle and you quote the name, like:

    select x as "FOO"

SQLAlchemy will see that as:

    select x as FOO

which in lowercase-as-case-insensitive translates to:

    select x as foo


Run this statement on PostgreSQL and Firebird:

    select 1 as foo

PostgresSQL cursor.description will show the name as "foo", Firebird will show it as "FOO".  who's correct?   Why wouldn't SQLAlchemy "normalize" that?

if you want to undo the name normalize, and have "select 1 as foo" return FOO on Oracle and Firebird, do a name denormalize for those backends.   Meaning, `name = name.upper() if name == name.lower()`.   the information is there in every case except a quoted all-uppercase name on Firebird / Oracle.

**Reply by fdcastel** [reply]
> PostgresSQL cursor.description will show the name as "foo", Firebird will show it as "FOO". who's correct?

Both. 

> Why wouldn't SQLAlchemy "normalize" that?

May I return the question and ask "Why it should?" 

This normalization is breaking what the user asked (a column explicitly named "foo" or "FOO").

If the user did not enforce the name with double quotes so it's ok to use whatever the RDBMS returns.
But if he DOES explicitly ask for a "foo" or "FOO" column, why SQLAlchemy should interfere and change this?

Also, I must insist, don't you see anything wrong with my original sample? Both SQL Server and Postgres have the expected behavior. Only Firebird (and Oracle) does not.

> if you want to undo the name normalize, and have "select 1 as foo" return FOO on Oracle and Firebird, do a name denormalize for those backends. Meaning, `name = name.upper() if name == name.lower()`.

This will not work if the user asked for a column named "all_lower" (column "three" in my original sample).

**Reply by CaselIT** [reply]
> May I return the question and ask "Why it should?

Because one of the objectives of sqlalchemy is to allow users to write a query once in its query language and run it on multiple database without needing to change the code.

If each database returns the column names in a different manner this is not possible.

The fact that as far as I know this is the first time this came up suggest that's a good approach.
That doesn't mean that there shouldn't be a way of turning off this behaviour, maybe allowing `requires_name_normalize` to be set in the dialect init

**Reply by zzzeek** [reply]
> 
> > Why wouldn't SQLAlchemy "normalize" that?
> 
> May I return the question and ask "Why it should?"

because in the vast majority of cases if Oracle tells us the column name was FOO the user actually asked for "foo" (without quotes), and we'd like these names to match that of more commonly used databases like SQLite, PostgreSQL, MySQL for the same SQL statement.  Using "QUOTED_ALL_UPPERCASE_NAMES" for SQL labels is not a standard practice, so that is the one case we let go for these lesser used databases.   cursor.description unfortunately does not tell us if a name was quoted or not if a name is given in a single case.

this directly impacts the names that we use for a named tuple:

      row = connection.execute(text("select 1 as foo")).first()
      print(f"the value is: {row.foo}")

Can you think of another way to achieve the above without name normalization ?
 
> 
> This will not work if the user asked for a column named "all_lower" (column "three" in my original sample).

I forgot to add that there is no *reason* to use name denormalize.   please continue to use cursor.description to get the raw data sent by the driver.

**Reply by gordthompson** [reply]
@fdcastel re: "This normalization is breaking what the user asked (a column explicitly named "foo" or "FOO")."

But *not* normalizing can also lead to a mis-match between what we ask for and what we get. Without normalization (by using the DBAPI connection directly):

```python
with engine.connect() as conn:
    dbapi_connection = conn.connection
    dbapi_cursor = dbapi_connection.cursor()
    try:
        dbapi_cursor.execute("DROP TABLE student")
    except:
        pass
    dbapi_cursor.execute(
        "CREATE TABLE student (id int primary key, firstname varchar(50))"
    )
    dbapi_connection.commit()
    dbapi_cursor.execute(
        "INSERT INTO student (id, firstname) VALUES (1, 'Gord')"
    )
    dbapi_connection.commit()
    dbapi_cursor.execute("SELECT firstname FROM student WHERE id = 1")
    print([row[0] for row in dbapi_cursor.description])
    # ['FIRSTNAME']
```

**Comment by fdcastel** [comment]
BTW: setting `requires_name_normalize` to `False` in Firebird dialects produces (what I believe) the correct output:

```python
def test_column_names_firebird():
    # FBDialect.requires_name_normalize = False
    e = sa.create_engine(db_uri)
    with e.connect() as conn:
        q = 'SELECT 1 one, 2 TWO, 3 "three", 4 "FOUR", 5 "Five" from rdb$database'
        r = conn.exec_driver_sql(q)
        assert list([x for x, *rest in r.cursor.description]) == ["ONE", "TWO", "three", "FOUR", "Five"]

        df = pd.read_sql(sql=q, con=db_uri, dtype_backend="pyarrow")
        assert list(df.columns.values) == ["ONE", "TWO", "three", "FOUR", "Five"]  # Correct. Same as r.cursor.description
```

Problem is: now the vast majority of SQLAlchemy tests are broken.

**Reply by fdcastel** [reply]
Please, I recognize that we are not in the same page here. But I'm trying to find the rationale for all this.

- Names enclosed in double quotes MUST be preserved exactly as declared.
- Names not enclosed in double quotes MAY be preserved (SQL Server);
  -  or converted to whatever the RDBMS prefers (lower for PostgreSQL, upper for Oracle/Firebird).

What's wrong with the above? If my memory serves me right every library out there follows these rules.

What I'm missing here? I'm surely not seeing something which you consider essential.

**Reply by CaselIT** [reply]
In a textual query how does sqlalchemy knowns if a name is in double quote or not?

**Reply by fdcastel** [reply]
Thank you guys, for your time and patience. I'm starting to see where this deadlock started.

> In a textual query how does sqlalchemy knowns if a name is in double quote or not?

@CaselIT My initial response would be: 

"It doesn't need to. If the user explicitly passed a TEXT SQL query (not one built by SQLAlchemy), it should be reasonable to assume that they know what they are asking for".

Also, 

> Because one of the objectives of sqlalchemy is to allow users to write a query once _**in its query language**_ and run it on multiple database without needing to change the code.

I fully agree with you. But please notice the highlight. As I see:

- (1) Queries written in SQLAlchemy query language _MUST_ be portable across databases/dialects;
- (2) Queries written in TEXT SQL _MAY_ not be. So;
  - (3) If the user chooses this path it should also accept the burden that comes with maintaining SQL queries across different databases.



But I started to see the root of the problem when @zzzeek said:

> ```python
> row = connection.execute(text("select 1 as foo")).first()
> print(f"the value is: {row.foo}")
> ``` 
> Can you think of another way to achieve the above without name normalization ?

@zzzeek My answer to this would be "do not use a textual query if you want it to be portable between different databases. Use a SQLAlchemy-built query. Or surround `foo` with double quotes.". 

As I see, the above example is clearly not portable. It _SHOULD_ return `1` in PostgreSQL and `None` in Oracle.

The "portable" way should be:

``` 
> row = connection.execute(select(literal_column("1").label("foo")).first()
> print(f"the value is: {row.foo}")
``` 

or even 

> ```python
> row = connection.execute(text('select 1 as "foo"')).first()
> print(f"the value is: {row.foo}")
> ``` 

but not as in the @zzzeek above sample.


Problem is that somewhere between my points (2) and (3) above, SQLAlchemy tried to choose a compromise solution. This may satisfy some "ORM" users. But certainly not "Core" users (my case).



Questions:

- Are my assumptions (1), (2) and (3) reasonable?
- Would this be a hard change to make now? Probably, yes. But It doesn't hurt to ask... 😅 

For now I will probably be taking the path suggested by @CaselIT: to pass a flag to disable `requires_name_normalize`.

**Reply by CaselIT** [reply]
Well text is part of the query language, like you can combine it freely with the other constructs.

But I think you have a point when using a string directly with `exec_driver_sql`. I'm not sure what it does now, but think not normalizing names then would make sense. Thought about this @zzzeek ?

**Reply by zzzeek** [reply]
> I fully agree with you. But please notice the highlight. As I see:
> 
>     * (1) Queries written in SQLAlchemy query language _MUST_ be portable across databases/dialects;
> 
>     * (2) Queries written in TEXT SQL _MAY_ not be. So;
> 
> @zzzeek My answer to this would be "do not use a textual query if you want it to be portable between different databases. Use a SQLAlchemy-built query. Or surround `foo` with double quotes.".

the change you propose means textual SQL queries would be fully incompatible across backends.    Our users would be completely opposed to this change.  it's not reasonable to require quotes for trivial, non-case-sensitive label names.  labels are best used in a case insensitive fashion and that's how SQLAlchemy is organized.


>       * (3) If the user chooses this path it should also accept the burden that comes with maintaining SQL queries across different databases.

for trivial textual SQL queries this would not be acceptable.

> Problem is that somewhere between my points (2) and (3) above, SQLAlchemy tried to choose a compromise solution. This may satisfy some "ORM" users. But certainly not "Core" users (my case).

it's been this way for....17 years.   first time anyone has ever said a thing about it.  so this is a horse that hasn't just left the barn, it's retired and had grand-horses.   sorry

I still dont really understand the original use case either.   I would suggest appealing to db-sig to add a new entry to cursor.description indicating case sensitivity of names.

> For now I will probably be taking the path suggested by @CaselIT: to pass a flag to disable `requires_name_normalize`.

this will definitely cause bugs for things like table creation and reflection - you will have to name your Table objects in UPPERCASE when using oracle or firebird and in lowercase on other backends.   

In addition to not understanding the use case I also dont understand what's wrong with using cursor.description for whatever it is you're trying to do.

**Reply by zzzeek** [reply]
just to add, this is not just about quoting of label names.  usually someone just selects the column names from a table.  

if you have a table named "user", and you select "id" and "name":

     result = connection.execute(text("select id, name from user"))

There is no opportunity for quoting there without placing the extra burden of explicitly labeling every column.  users will have code that runs fine on SQLite, PostgreSQL, Mysql for years, then when they go to add Oracle support, all such queries are broken.  This would not be acceptable either.

Also note the "name normalize" could easily be bypassed by adding an execution option for "disable_name_normalize=True".   then you wouldn't be introducing bugs in schema management.

**Reply by CaselIT** [reply]
> But I think you have a point when using a string directly with exec_driver_sql. I'm not sure what it does now, but think not normalizing names then would make sense. Thought about this @zzzeek ?

`exec_driver_sql` currently applies the name normalization, like the other methods. 
A case could be made by turning this off in this case, but I'm not sure how popular would be this option

**Reply by fdcastel** [reply]
While this may be a good option to add (I can't see any cons), it probably will not solve my problem because Pandas uses `execute()` in most of its code.

It was just a quick glimpse, but I found only [one](https://github.com/pandas-dev/pandas/blob/main/pandas/io/sql.py#L1654) usage of `exec_driver_sql` in pandas source code. All others are `execute()`.

**Reply by zzzeek** [reply]
please note my comment at https://github.com/sqlalchemy/sqlalchemy/discussions/10781#discussioncomment-7911481:

> Also note the "name normalize" could easily be bypassed by adding an execution option for "disable_name_normalize=True". then you wouldn't be introducing bugs in schema management.

[documentation for execution options](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Connection.execution_options)

**Reply by fdcastel** [reply]
Working on a POC for this now (passing `disable_name_normalize` in `execution_options`)

**Reply by fdcastel** [reply]
Uh. How could I access `execution_options` from my `FBDialect` class?

Both `normalize_name` and `denormalize_name` are in this class. But, other than the name to be converted, they don't receive any other information (connection, context, etc.) about the current execution.

```python
class FBDialect(default.DefaultDialect):
    # ...
    def normalize_name(self, name):
        execution_options = self.<HERE-THE-MAGIC-HAPPENS>.get_execution_options()
        if execution_options.get("disable_name_normalize", False):
            return name
        return super().normalize_name(name)

    def denormalize_name(self, name):
        execution_options = self.<HERE-THE-MAGIC-HAPPENS>.get_execution_options()
        if execution_options.get("disable_name_normalize", False):
            return name
        return super().denormalize_name(name)
```

In other words: I'm missing the magic part. 😅

**Reply by fdcastel** [reply]
Well... It seems that thanks to an [old SQLite hack](https://github.com/sqlalchemy/sqlalchemy/blob/99da5ebab36da61b7bfa0b868f50974d6a4c4655/lib/sqlalchemy/engine/cursor.py#L483) (and a little help from the ingenious `quoted_name` class) I can pass the information needed from `FBExecutionContext` to `normalize_name`/`denormalize_name` dialect functions:

```python
class FBExecutionContext(default.DefaultExecutionContext):
    # ...
    @util.memoized_property
    def _disable_name_normalize(self):
        return self.execution_options.get("disable_name_normalize", False)

    def _translate_colname(self, colname):
        if self._disable_name_normalize:
            return quoted_name(colname, quote=True), colname

        return colname, None

class FBDialect(default.DefaultDialect):
    # ...
    def normalize_name(self, name):
        if isinstance(name, quoted_name) and name.quote:
            return name

        return super().normalize_name(name)

    def denormalize_name(self, name):
        if isinstance(name, quoted_name) and name.quote:
            return name

        return super().denormalize_name(name)
```

Tests passing:

```python
def test_column_names_are_normalized(self, connection):
    q = 'SELECT 1 one, 2 TWO, 3 "three", 4 "FOUR", 5 "Five" from rdb$database'
    r = connection.exec_driver_sql(q)

    original = ["ONE", "TWO", "three", "FOUR", "Five"]
    assert [c for c, *rest in r.cursor.description] == original

    normalized = ["one", "two", "three", "four", "Five"]
    assert [c for c in r.mappings().first()] == normalized

def test_column_names_normalization_can_be_disabled(self, connection):
    connection.execution_options(disable_name_normalize=True)

    q = 'SELECT 1 one, 2 TWO, 3 "three", 4 "FOUR", 5 "Five" from rdb$database'
    r = connection.exec_driver_sql(q)

    original = ["ONE", "TWO", "three", "FOUR", "Five"]
    assert [c for c, *rest in r.cursor.description] == original
    assert [c for c in r.mappings().first()] == original
```

And... Yes. I can understand that the last thing you want to see now is another crazy guy wanting to hook another code into that `_translate_colname`. Opinions are welcome! 😉

**Reply by zzzeek** [reply]
my proposal was to accomodate this execution option [here](https://github.com/sqlalchemy/sqlalchemy/blob/99da5ebab36da61b7bfa0b868f50974d6a4c4655/lib/sqlalchemy/engine/cursor.py#L659).   we are talking about a new feature

**Reply by fdcastel** [reply]
Understood. That would be my next question. 

This would be useful for any other dialects using `requires_name_normalize = True`. So it could be implemented in the SQLAlchemy itself.

If you wish, I can work on some PRs for this.

**Reply by zzzeek** [reply]
this has a lot of twists and turns to it, not the least of which is the `ResultMetaData` structure is normally cached which breaks things badly here if not disabled, so stay tuned...

**Reply by zzzeek** [reply, marked as answer]
the issue is at #10789.   this is a very trivial thing for exec_driver_sql and text() without any columns.  however for everything else, when Column objects are present, it's a gigantic change to CursorResult to do this while keeping type handlers working, so this is a 2.1 only thing at best.  you'll have to hardcode workarounds in the meantime.

**Reply by fdcastel** [reply]
Great! Thank you very much @zzzeek!

I can keep using my hacks until 2.1 release. 😄 

Thanks, guys (@zzzeek @CaselIT @gordthompson )! You're awesome!

**Comment by fdcastel** [comment]
> the change you propose means textual SQL queries would be fully incompatible across backends. Our users would be completely opposed to this change.

Fully comprehensive. 



> I still don't really understand the original use case either. 

I need to run this query (via Pandas):

`SELECT 3 "three", 4 "FOUR", 5 "Five"`

from 3 different databases. And merge their results (data frames).

Just looking to the original query, there is no ambiguity about what the name of columns should be.  However

- SQL Server returns `["three", "FOUR", "Five"]`
- PostgreSQL returns `["three", "FOUR", "Five"]`
- Firebird returns `["three", "four", "Five"]`

Since Firebird does not return the column name that was asked for the merge fails. This discrepancy across dialects is what started the discussion. 



> > to pass a flag to disable `requires_name_normalize`.

> this will definitely cause bugs for things like table creation and reflection

I just need to read data. In this particular case Pandas use SQLAlchemy only as a low-level data access library for a database-agnostic way.

That's also related with my "ORM" x "Core" users I mentioned before. At lower levels, any normalization is a burden.



> I also don't understand what's wrong with using `cursor.description` for whatever it is you're trying to do.

Since the hard work here is being done by Pandas, I don't have an opportunity to touch `cursor.description` nor change the generated columns. 

I need to run the query a second time (probably getting only its metadata, e.g. with a `ROWS 0`). Not ideal. But, also, not a big problem in this case. -- _BTW: This was the original problem for #10771_

**Reply by zzzeek** [reply]
yah, quoted UPPERCASE names on name-denormalized DBs is the use case we have never supported.   I would advise against using case sensitive label names like that, these are uncommon in relational database use and only make things more difficult.

**Reply by zzzeek** [reply]
> I need to run this query (via Pandas):
> 
> `SELECT 3 "three", 4 "FOUR", 5 "Five"`
> 
> from 3 different databases. And merge their results (data frames).
> 
> Just looking to the original query, there is no ambiguity about what the name of columns should be. However
> 
>     * SQL Server returns `["three", "FOUR", "Five"]`
> 
>     * PostgreSQL returns `["three", "FOUR", "Five"]`
> 
>     * Firebird returns `["three", "four", "Five"]`

if the use case is just one of merging results, then:

1.  if the same query is run on every database, you dont need the names to merge the results, do it positionally
2. to merge on name, the names are unique when taken case insensitively, you can merge on name.lower()

if you *dont* have control of the query, then you dont know if the label names are quoted or not which means if unquoted "three" and unquoted "FOUR" are used, then you cannot possibly get a consistent result on all backends.

**Reply by fdcastel** [reply]
I don't have query control. That's why I told my (SQL-aware) users that they must keep their names consistent across different databases... USING DOUBLE QUOTES AROUND COLUMN NAMES! 😄

Lowering everything before merge it's also not possible. I really need to preserve the original names they used. Because these column names would be needed later in other processes (scripts).

Also, it fails for`["three", "four", "FOUR"]` case.

Later, my users would want to access the values using the same names they provided in their input queries. e.g.:

```
merged_values['three']
merged_values['four']
merged_values['FOUR']
```

**Reply by zzzeek** [reply]
> Hi! I recently encountered same problem with quoted UPPERCASE names in an Oracle database. I acknowledge that it's generally better to avoid using case-sensitive identifiers, but it's a common scenario where you don't have control over the data (for instance, when your product is integrated with databases owned by other companies).

it's not a common case since we've never really gotten reports of it until this issue, but yes, in order to try to be database agnostic we had to decide on something that would work in the most cases possible and it was this one.

> 
> Therefore, I'd like to suggest a solution for this issue.

well what issue?  The poster here is building something that has some degree of third party clients using any kind of name and expecting things to come back identically in dictionaries, which is not really SQLAlchemy's main focus.   you can use quoted uppercase names with Oracle there are just some caveats.


> What if the `normalize_name` function could accept an optional list of all names within the current normalization context (since names are typically normalized in "batches" of entities of the same type, e.g. columns within same table)? We could then apply the following rule for normalization:
> 
>     * If the name is uppercase and there exists another identical name in lowercase, do not normalize it.
> 
>     * Otherwise, apply the existing normalization rules.

you can already indicate names that are case sensitive using the [quoted_name](https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.quoted_name) construct, this doesn't apply quite as well to reflection since we are getting those names back from elsewhere, but there are ways to affect it on that end too using the [column_reflect](https://docs.sqlalchemy.org/en/20/core/events.html#sqlalchemy.events.DDLEvents.column_reflect) event hook.

 
> I understand that this is a very specific case, but I've found a few people in this conversation, including myself, who are dealing with such an issue. So, I hope this could be helpful.

I need to see an exact short demonstration program of the real-world use case that is what it is you actually need to achieve.    not clear if we are dealing with reflection here, or another odd reliance on the names in cursor.description with raw textual queries for which my proposal is at #10789

**Reply by CaselIT** [reply]
how wired the comment you replied to was deleted

**Reply by fdcastel** [reply]
Just to chime in: From my side, the solution proposed by @zzzeek in #10789 is perfect and it would serve me.

Waiting anxiously for 2.1 😉
